File: backend/app/api/v1/endpoints/tasks.py
"""
Task management endpoints.

This module provides CRUD endpoints for todo tasks with zero-trust security.
All operations are scoped to the authenticated user.
"""

from typing import List
from fastapi import APIRouter, Depends, HTTPException, status
from sqlmodel import Session, select

from app.auth.dependencies import get_current_user
from app.database import get_db
from app.models.task import Task
from app.schemas.task import (
    TaskResponse,
    TaskListResponse,
    TaskCreate,
    TaskUpdate,
)


router = APIRouter()


@router.get("", response_model=TaskListResponse)
async def list_tasks(
    current_user_id: str = Depends(get_current_user),
    session: Session = Depends(get_db),
) -> TaskListResponse:
    """
    List all tasks for the authenticated user.

    Returns tasks ordered by creation date (newest first).
    Implements zero-trust security - only returns tasks belonging to the authenticated user.

    Args:
        current_user_id: User ID from JWT token (injected by get_current_user)
        session: Database session

    Returns:
        TaskListResponse with tasks array and count
    """
    # Query tasks filtered by user_id and ordered by created_at DESC
    statement = (
        select(Task)
        .where(Task.user_id == current_user_id)
        .order_by(Task.created_at.desc())
    )
    tasks = session.exec(statement).all()

    return TaskListResponse(
        tasks=[TaskResponse.model_validate(task) for task in tasks],
        count=len(tasks),
    )


@router.post("", response_model=TaskResponse, status_code=status.HTTP_201_CREATED)
async def create_task(
    task_data: TaskCreate,
    current_user_id: str = Depends(get_current_user),
    session: Session = Depends(get_db),
) -> TaskResponse:
    """
    Create a new task for the authenticated user.

    The task is automatically associated with the authenticated user and marked as incomplete.

    Args:
        task_data: Task creation data (title, optional description)
        current_user_id: User ID from JWT token (injected by get_current_user)
        session: Database session

    Returns:
        Created task with 201 status code
    """
    # Generate a new UUID for the task
    import uuid
    task_id = str(uuid.uuid4())

    # Create task instance
    task = Task(
        id=task_id,
        title=task_data.title,
        description=task_data.description,
        completed=False,
        user_id=current_user_id,
    )

    # Save to database
    session.add(task)
    session.commit()
    session.refresh(task)

    return TaskResponse.model_validate(task)


@router.get("/{task_id}", response_model=TaskResponse)
async def get_task(
    task_id: str,
    current_user_id: str = Depends(get_current_user),
    session: Session = Depends(get_db),
) -> TaskResponse:
    """
    Get a specific task by ID.

    Implements zero-trust security - only returns the task if it belongs to the authenticated user.

    Args:
        task_id: Task ID from URL path
        current_user_id: User ID from JWT token (injected by get_current_user)
        session: Database session

    Returns:
        Task details

    Raises:
        HTTPException: 404 if task not found or doesn't belong to user
    """
    # Query task with user_id filter (zero-trust security)
    statement = select(Task).where(
        Task.id == task_id,
        Task.user_id == current_user_id,
    )
    task = session.exec(statement).first()

    if not task:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found",
        )

    return TaskResponse.model_validate(task)


@router.patch("/{task_id}", response_model=TaskResponse)
async def update_task(
    task_id: str,
    task_data: TaskUpdate,
    current_user_id: str = Depends(get_current_user),
    session: Session = Depends(get_db),
) -> TaskResponse:
    """
    Update a task (title, description, or completion status).

    Implements zero-trust security - only updates the task if it belongs to the authenticated user.

    Args:
        task_id: Task ID from URL path
        task_data: Task update data (optional title, description, completed)
        current_user_id: User ID from JWT token (injected by get_current_user)
        session: Database session

    Returns:
        Updated task

    Raises:
        HTTPException: 404 if task not found or doesn't belong to user
    """
    # Query task with user_id filter (zero-trust security)
    statement = select(Task).where(
        Task.id == task_id,
        Task.user_id == current_user_id,
    )
    task = session.exec(statement).first()

    if not task:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found",
        )

    # Apply partial update (only update provided fields)
    if task_data.title is not None:
        task.title = task_data.title
    if task_data.description is not None:
        task.description = task_data.description
    if task_data.completed is not None:
        task.completed = task_data.completed

    # Save changes
    session.add(task)
    session.commit()
    session.refresh(task)

    return TaskResponse.model_validate(task)


@router.delete("/{task_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_task(
    task_id: str,
    current_user_id: str = Depends(get_current_user),
    session: Session = Depends(get_db),
) -> None:
    """
    Delete a task permanently.

    Implements zero-trust security - only deletes the task if it belongs to the authenticated user.

    Args:
        task_id: Task ID from URL path
        current_user_id: User ID from JWT token (injected by get_current_user)
        session: Database session

    Returns:
        204 No Content on success

    Raises:
        HTTPException: 404 if task not found or doesn't belong to user
    """
    # Query task with user_id filter (zero-trust security)
    statement = select(Task).where(
        Task.id == task_id,
        Task.user_id == current_user_id,
    )
    task = session.exec(statement).first()

    if not task:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found",
        )

    # Delete task
    session.delete(task)
    session.commit()

File: frontend/lib/api/tasks.ts
/**
 * Task API client functions.
 *
 * All functions automatically include session token from cookies.
 */

import { Task, TaskListResponse, CreateTaskRequest, UpdateTaskRequest } from "@/types/task";

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

/**
 * Get session token from cookies.
 */
function getSessionToken(): string | null {
  if (typeof document === "undefined") return null;

  const cookies = document.cookie.split(";");
  for (const cookie of cookies) {
    const [name, value] = cookie.trim().split("=");
    if (name === "better-auth.session_data") {
      return decodeURIComponent(value);
    }
  }
  return null;
}

/**
 * List all tasks for the authenticated user.
 */
export async function listTasks(): Promise<TaskListResponse> {
  const token = getSessionToken();

  const response = await fetch(`${API_BASE_URL}/api/v1/tasks`, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
  });

  if (!response.ok) {
    if (response.status === 401) {
      throw new Error("Not authenticated");
    }
    throw new Error(`Failed to fetch tasks: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Create a new task.
 */
export async function createTask(data: CreateTaskRequest): Promise<Task> {
  const token = getSessionToken();

  const response = await fetch(`${API_BASE_URL}/api/v1/tasks`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
    body: JSON.stringify(data),
  });

  if (!response.ok) {
    if (response.status === 401) {
      throw new Error("Not authenticated");
    }
    if (response.status === 400) {
      const error = await response.json();
      throw new Error(error.detail || "Invalid task data");
    }
    throw new Error(`Failed to create task: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Get a specific task by ID.
 */
export async function getTask(taskId: string): Promise<Task> {
  const token = getSessionToken();

  const response = await fetch(`${API_BASE_URL}/api/v1/tasks/${taskId}`, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
  });

  if (!response.ok) {
    if (response.status === 401) {
      throw new Error("Not authenticated");
    }
    if (response.status === 404) {
      throw new Error("Task not found");
    }
    throw new Error(`Failed to fetch task: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Update a task.
 */
export async function updateTask(taskId: string, data: UpdateTaskRequest): Promise<Task> {
  const token = getSessionToken();

  const response = await fetch(`${API_BASE_URL}/api/v1/tasks/${taskId}`, {
    method: "PATCH",
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
    body: JSON.stringify(data),
  });

  if (!response.ok) {
    if (response.status === 401) {
      throw new Error("Not authenticated");
    }
    if (response.status === 404) {
      throw new Error("Task not found");
    }
    if (response.status === 400) {
      const error = await response.json();
      throw new Error(error.detail || "Invalid task data");
    }
    throw new Error(`Failed to update task: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Delete a task.
 */
export async function deleteTask(taskId: string): Promise<void> {
  const token = getSessionToken();

  const response = await fetch(`${API_BASE_URL}/api/v1/tasks/${taskId}`, {
    method: "DELETE",
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
  });

  if (!response.ok) {
    if (response.status === 401) {
      throw new Error("Not authenticated");
    }
    if (response.status === 404) {
      throw new Error("Task not found");
    }
    throw new Error(`Failed to delete task: ${response.statusText}`);
  }
}
